pkg/component/component.go

```diff
@@ -367,104 +387,144 @@ func GetComponentSource(client *occlient.Client, componentName string, applicati
+		glog.V(4).Infof("Updating the DeploymentConfig %s image to %s", namespacedOpenShiftObject, bc.Spec.Output.To.Name)
+		err = client.UpdateDCToGit(commonObjectMeta, bc.Spec.Output.To.Name)
+		if err != nil {
+			return errors.Wrapf(err, "unable to update DeploymentConfig image for %s component", componentName)
+		}
+
+		// Cleanup after the supervisor
 		err = client.CleanupAfterSupervisor(namespacedOpenShiftObject, projectName, annotations)
 		if err != nil {
 			return errors.Wrapf(err, "unable to update DeploymentConfig  for %s component", componentName)
 		}
+
+		// Finally, we build!
 		err = Build(client, componentName, applicationName, true, true, stdout)
 		if err != nil {
 			return errors.Wrapf(err, "unable to build the component %v", componentName)
 		}
+
 	} else if oldSourceType == "git" && (newSourceType == "binary" || newSourceType == "local") {
 		// Steps to update component from git to local or binary
-		// - update odo annotations and labels to reflect changes
-		// - update BuildConfig source to have s2i supervisor bootstrap repository as a source https://github.com/kadel/bootstrap-supervisored-s2i
-		// - trigger build (wait for it to finish)
-		// - update and add supervisor to dc
-
-		labels := componentlabels.GetLabels(componentName, applicationName, true)
 
+		// Update the sourceURL since it is not a local/binary file.
 		sourceURL := util.GenFileUrl(newSource, runtime.GOOS)
-		annotations := map[string]string{componentSourceURLAnnotation: sourceURL}
-		annotations[componentSourceTypeAnnotation] = newSourceType
+		annotations[componentSourceURLAnnotation] = sourceURL
 
-		err = client.UpdateBuildConfig(namespacedOpenShiftObject, projectName, "", annotations)
+		// Need to delete the old BuildConfig
+		err = client.DeleteBuildConfig(commonObjectMeta)
 		if err != nil {
-			return errors.Wrapf(err, "unable to update BuildConfig  for %s component", componentName)
-		}
-		err := Build(client, componentName, applicationName, false, true, stdout)
-		if err != nil {
-			return errors.Wrapf(err, "unable to build the component %v", componentName)
+			return errors.Wrapf(err, "unable to delete BuildConfig for %s component", componentName)
 		}
 
-		err = client.SetupForSupervisor(namespacedOpenShiftObject, projectName, annotations, labels)
+		// Update the DeploymentConfig
+		err = client.UpdateDCToSupervisor(commonObjectMeta, componentImageType)
 		if err != nil {
-			return errors.Wrapf(err, "unable to update DeploymentConfig  for %s component", componentName)
+			return errors.Wrapf(err, "unable to update DeploymentConfig for %s component", componentName)
 		}
+
 	} else {
 		// save source path as annotation
 		// this part is for updates where the source does not change or change from local to binary and vice versa
 
-		var annotations map[string]string
 		if newSourceType == "git" {
-			annotations = map[string]string{componentSourceURLAnnotation: newSource}
```

Comment:
Don't we need this anymore?
Bcoz the comment says this can be also be the case "change from local to binary and vice versa" apart from the other possibility "updates where the source does not change".

Reply:
Nope. We do this higher up in the code.

Reply:
ok so you mean here : https://github.com/redhat-developer/odo/pull/622/files#diff-692effa1b0b7a9ed8c1f7bb49a977b9aR412